src/db.ts

```diff
@@ -918,33 +917,40 @@ export abstract class Database {
           continue;
         }
 
-        updateJobs.recorded(values[1], values[0], values[4]);
-        batchedMs += 1;
-        batchedValues = batchedValues.concat(values);
-        if (batchedMs === Database.batchN) {
-          try {
-            const result = await this.recordMeasurementBatchedN(batchedValues);
-            recordedMeasurements += result;
-          } catch (err) {
-            // we may have concurrent inserts, or partially inserted data,
-            // where a request aborted
-            if (isUniqueViolationError(err)) {
-              recordedMeasurements += await this.recordMeasurementsFromBatch(
-                batchedValues
-              );
-            } else {
-              throw err;
-            }
+        let iMID = '';
+        // build a id for each unquie Measurement
+        iMID = iMID.concat(
+          run.id.toString() +
+            ' ' +
+            trial.id.toString() +
+            ' ' +
+            d.in.toString() +
+            ' ' +
+            criteria.get(m.c).id.toString()
+        );
+        if (iterationMap.has(iMID)) {
+          let value = iterationMap.get(iMID);
+          if (value !== undefined) {
+            value = value.concat([m.v]);
+          } else {
+            throw new Error('value is undefined');
           }
-          batchedValues = [];
-          batchedMs = 0;
+          iterationMap.set(iMID, value);
+        } else {
+          iterationMap.set(iMID, [m.v]);
         }
+
+        updateJobs.recorded(values[1], values[0], values[4]);
       }
     }
+    // build list of batchedValues
+    iterationMap.forEach((value: Array<number>, key: string) => {
+      batchedValues = batchedValues.concat(key.split(' ').map(Number), [value]);
```

Comment:
Could you assign the result of `key.split(' ').map(Number)` in a well-named variable first to document implicitly what happens here?